### ISSUE TITLE
[Feature] Bulk upgrade/evolve of Pokemon

### ISSUE THREAD

**Issue by Senney** [cglatot/PokeManager#28]
Great app, loving it so far!

I've started implementing this myself, so maybe I can submit it as a pull request once I've got it complete.

It would be great to be able to quickly Upgrade a specific Pokemon a specified number of times, or until you run out of candies/dust. Same idea with Evolving, except evolve all Pokemon of a single type (useful for cases where you're farming XP or have popped an egg). I've started work on this, and it seems to be doable using the UpgradePokemonMessage_pb2 and EvolvePokemonMessage_pb2 messages.

I'll submit a Pull Request some time this weekend or early next week with my implementation.

**Comment by cglatot** [comment]
You shouldn't need to go to the protobufs at all, there are methods in Session that allow you to evolve. Look at the API for info. And look at their demo.py - it has evolve demos in there: https://github.com/rubenvereecken/pokemongo-api

If you are going to submit a pull request for this make sure you do the following:
- Check for number of candies and calculate how many evolutions can happen for that pokemon (and display to user)
- Correctly deduct candies when evolving (or make sure that you don't try to evolve more than can be evolved)
- Allow the user to specify how many to evolve for each pokemon
- Allow the user to specify a certain Pokemon (and how many to evolve), or to evolve everything available (or at the very least, code it in a way that the second feature could be added on easily enough)

If you meet those criteria, and it follows similar formatting / tabular layout as the rest I'll be happy to merge. Looking at the transfer and counts methods will likely provide most of the code for it anyway.

**Comment by fprovostpoulin** [comment]
Is there a way to ask to activate a lucky egg before starting the evolutions ? That way you can as well tell if you have the optimal number of upgrades.

Also, it may be nice to have an option to filter by How much candy it requires... that way we could focus on just pidgey, caterpies and weedles if we want (Or as it'S done in the "Transfer" option, maybe ask for what pokemon we want to evolve, and how many of them starting by the higher IV maybe )

**Comment by MatyiFKBT** [comment]
@fprovostpoulin maybe you need to activate the egg on your phone and than run the code on your computer. If you don't happen to have thousands of Pokémons, it should be done in 30 minutes.

**Comment by fprovostpoulin** [comment]
can't we use session.evolvePokemon(monster[6]) in order to do it ? I didn't succeed to evolve through this method. Any help ?

**Comment by BeardAnnihilator** [comment]
#59
